### PR TITLE
Refactor monitoring content to improve SapHanaCheck

### PR DESCRIPTION
**Pull request by tniek**
- Previously, a lot of logic to schedule checks and validate connectivity was in the main module (sapmon.py). This limited the functionality of the monitoring and onboarding payload.
- This PR refactors the monitoring content into classes in its own right. On a high level, this has resulted in the following improvements:
  - Every content provider has now its own state file, which improves readability.
  - Any monitoring check can now have a list of actions that are executed when the check is run; each action can have its own set of parameters.
  - Every content provider has now a validation method that is run during onboarding.
- Specifically for SAP HANA (and the monitoring checks of type `SapHanaCheck`), the above were pre-requisites for the following features:
  - Client-side failover logic for scale-out (multi-host) landscapes is now homegrown, since SAP's official hdbcli Python client does provide native support.
  - SqlProbe check continuously monitors if a SQL connection can be established to each of the HANA nodes and measures the latency.

### CODE REVIEW

File: sapmon/payload/provider/saphana.py [hunk 1]
```diff
@@ -14,180 +15,391 @@
 
 ###############################################################################
 
-# Provide access to a HANA Database (HDB) instance
-class SapHana:
-   TIMEOUT_HANA_SECS = 5
+# Stores configuration specific to a SAP HANA database instance
+class SapHanaConfig(metaclass=Singleton):
```
Comment (persiaAziz) [inline sapmon/payload/provider/saphana.py]: Why is this being made a singleton? This will need to be refactored when we move to multi-instance, no?

Comment (tniek) [reply sapmon/payload/provider/saphana.py]: Yes, absolutely. See my comment in sapmon.py:123
 `# Until we have multiple provider instances, just pick the first HANA config`

Implementing both multiple provider types (besides SAP HANA) and allowing multiple instances of providers (e.g. 2 HANA + 1 Pacemaker) will require a lot of refactoring.

File: sapmon/payload/provider/base.py [hunk 2]
```diff
@@ -1,44 +1,168 @@
 # Python modules
 from abc import ABC, abstractmethod
+from datetime import date, datetime, timedelta
 import logging
+import json
+from typing import Callable, Dict, List, Optional
+
+# Payload modules
+from const import *
+from helper.tools import *
```
Comment (rukawata) [inline sapmon/payload/provider/base.py]: Can we sort the from import and import?

File: sapmon/payload/sapmon.py [hunk 3]
```diff
@@ -80,94 +80,19 @@ def initMonitoringContent(self) -> None:
...
-         contentVersion = jsonData.get("contentVersion", None)
-         if not contentVersion:
-            appTracer.error("content version not specified in content file %s, skipping" % contentFullPath)
-            continue
-         checks = jsonData.get("checks", [])
-         if not contentType in sapmonContentTypes:
-            appTracer.error("unknown content type %s, skipping content file %s" % (contentType, contentFullPath))
-            continue
+         #providerClass = getattr(self, providerName)
+         #contentProvider = providerClass(appTracer, contentFullPath)
```
Comment (rukawata) [inline sapmon/payload/sapmon.py]: If we no longer need those lines, please remove.

File: sapmon/payload/sapmon.py [hunk 4]
```diff
@@ -234,6 +150,20 @@ def sliceDict(d: dict, s: str) -> dict:
...
+                               resultJson: str) -> None:
+      appTracer.info("sending customer analytics")
+      results = json.loads(resultJson, object_hook=JsonDecoder.datetimeHook)
+      for result in results:
+         metrics = {
+            "Type": c.customLog,
+            "Data": result,
+         }
+         appTracer.debug("metrics=%s" % metrics)
+         analyticsTracer.info(metrics)
```
Comment (PakDLiu) [inline sapmon/payload/sapmon.py]: I have a PR that will update this slightly: https://github.com/Azure/AzureMonitorForSAPSolutions/pull/14

**Pull request merged**